File: pbmx_ffi/java/io/rmf/pbmx/Random.java
package io.rmf.pbmx;

import io.rmf.pbmx.ffi.LibPbmx;
import io.rmf.pbmx.ffi.RawRng;
import com.sun.jna.ptr.LongByReference;

public final class Random {
    public static long[] permutation(int length) {
        return Random.doPermutation(length, null);
    }

    public static long[] permutation(int length, Rng rng) {
        return Random.doPermutation(length, RawRng.wrap(rng));
    }

    private static long[] doPermutation(int length, RawRng rng) {
        long[] p = new long[length];
        int r = LibPbmx.INSTANCE.pbmx_random_permutation(rng, length, p);
        assert r != 0;

        return p;
    }

    public static int shift(int length) {
        return Random.doShift(length, null);
    }

    public static int shift(int length, Rng rng) {
        return Random.doShift(length, RawRng.wrap(rng));
    }

    private static int doShift(int length, RawRng rng) {
        LongByReference outK = new LongByReference();
        int r = LibPbmx.INSTANCE.pbmx_random_shift(rng, length, outK);
        assert r != 0;

        return (int)outK.getValue();
    }
}

File: pbmx_droid/app/src/main/java/io/rmf/pbmx/Random.kt
package io.rmf.pbmx

import com.sun.jna.Pointer
import com.sun.jna.ptr.LongByReference
import io.rmf.pbmx.ffi.FFI

fun randomPermutation(length: Int): LongArray {
    val p = LongArray(length)
    val r = FFI.pbmx_random_permutation(Pointer.NULL, length.toLong(), p)
    assert(r != 0)

    return p
}

fun randomShift(length: Int): Int {
    val outK = LongByReference()
    val r = FFI.pbmx_random_shift(Pointer.NULL, length.toLong(), outK)
    assert(r != 0)

    return outK.value.toInt()
}
